netgrasp/utils/mac.py

```diff
@@ -0,0 +1,255 @@
+DB_LOOKUP_CONFIG_UPDATED_KEY = 'maclookup_updated'
+
+LOOKUP_SOURCES = ['local', 'remote']
+
+DEFAULT_LOOKUP_SOURCE = 'remote'
+DEFAULT_LOOKUP_UPDATED = datetime.datetime(1970, 1, 1)
+DEFAULT_LOOKUP_CONFIG = dict(source=DEFAULT_LOOKUP_SOURCE, updated=DEFAULT_LOOKUP_UPDATED)
+
+WIRESHARK_ATTRIBUTION = """\
+data provided by: Wireshark - Network traffic analyzer
+                  Copyright 1998 Gerald Combs
+                  Wireshark is free software; licensed under the terms of
+                  the GNU General Public License.\
+"""
+
+
+def normalize_mac(mac):
+    """Return upper-case mac with leading 0s and ':' delimiter."""
+    mac = mac.upper()
+    pieces = re.split('[-:.]', mac)
+    mac = ':'.join(pieces)
+    if not re.match('[0-9A-F]{2}([:])[0-9A-F]{2}(\\1[0-9A-F]{2}){4}$', mac):
+        fixed = []
+        pieces = mac.split(':')
+        for piece in pieces:
+            if len(piece) == 1:
+                piece = '0' + piece
+            fixed.append(piece)
+        mac = ':'.join(fixed)
+    return mac
+
+
+def lookup_local(mac, dbfile, lockfile, debugger):
+    """Return vendor or None if not found. Raise if error."""
+    from netgrasp.database import database
+
+    debugger.debugger.debug('entering mac.lookup_local(%s)' % mac)
+
+    mac = normalize_mac(mac)
+    debugger.debug('looking up vendor for %s' % mac)
+    # Lookup vendor in db. Match mac using netmask if > 0.
+    # @TODO matching needs to be implemented; this is "placeholder" code
+    ##db = database.Database(dbfile, lockfile, debugger)
+    ##db.cursor.execute("SELECT * FROM %s WHERE ... ?" % DB_LOOKUP_TABLE_NAME, (mac,))
+    ##result = db.cursor.fetchall()
+    match = False
+    if match:
+        vendor = 'vendor'
+        debugger.debugger.debug('identified %s as %s' % (mac, vendor))
+    else:
+        vendor = None
+        debugger.debugger.info('failed to identify %s' % mac)
+
+    return vendor
+
+
+def lookup_remote(mac, debugger):
+    """Return vendor or None if not found. Raise if error."""
+    import httplib
+
+    debugger.debug('entering mac.lookup_remote(%s)' % mac)
+
+    mac = normalize_mac(mac)
+    debugger.debug('looking up vendor for %s' % mac)
+    http = httplib.HTTPConnection('api.macvendors.com', port=80)
+    url = '/%s' % mac
+    http.request('GET', url)
+    response = http.getresponse()
+    if response.status == httplib.OK:
+        vendor = response.read()
+        debugger.debugger.debug('identified %s as %s' % (mac, vendor))
+    else:
+        vendor = None
+        debugger.debugger.info('failed to identify %s' % mac)
+
+    return vendor
+
+
+def print_lookup_config(dbfile, lockfile, debugger):
+    ok, config = get_lookup_config(dbfile, lockfile, debugger)
+    if not ok:
+        print(config)
+        return
+    print('mac lookup configuration')
+    print('source: %s' % config['source'])
+    print('mac database was last updated: %s' % get_dbupdated_text(config['updated']))
```

Comment:
I'd rather more proactively help people keep things up-to-date: perhaps a configuration option for how often to auto-update the mac addresses (in days so people don't abuse the upstream and auto-update too often, perhaps default to every 7 days). This would auto-fire if "auto" or "local" are set, not if "remote" is set.